### PR TITLE
genChange impl for random scheme

### PR DESCRIPTION
**Pull request by paweljakubas**
# Issue Number

#560 

# Overview

- [x] I have extended RndState for ChangeState
- [x] I have added preliminary instance of GenChange
- [x] I have added property tests
- [x] I have added DB related stuff

### CODE REVIEW

File: lib/core/src/Cardano/Wallet/DB/Sqlite.hs [hunk 1]
```diff
@@ -906,7 +906,9 @@ instance PersistState W.RndState where
     deleteState _ = pure ()
 
     -- Construct a 'RndState' from the root private key
-    selectState (wid, _) = fmap (W.RndState . fst) <$> selectPrivateKey wid
+    selectState (wid, _) = do
+        let constructRndState key = W.RndState key W.emptyChangeState
```
Comment (KtorZ) [inline lib/core/src/Cardano/Wallet/DB/Sqlite.hs]: I suppose this is temporary, right :sweat_smile: ? We do need to store / retrieve the whole state into and from the database.

File: lib/core/src/Cardano/Wallet/Primitive/AddressDiscovery/Random.hs [hunk 2]
```diff
@@ -43,23 +42,31 @@ import Control.Monad
...
 
 instance NFData RndState
 
 -- An address is considered to belong to the 'RndState' wallet if it can be decoded
 -- as a Byron HD random address, and where the wallet key can be used to decrypt
 -- the address derivation path.
 instance IsOurs RndState where
-    isOurs addr st@(RndState key) =
+    isOurs addr st@(RndState key _) =
         (isJust $ addressToPath addr key, st)
```
Comment (KtorZ) [inline lib/core/src/Cardano/Wallet/Primitive/AddressDiscovery/Random.hs]: Note that addresses discovered by `IsOurs` should be added to the state. This is where we get to know which addresses are already known.

File: lib/core/src/Cardano/Wallet/Primitive/AddressDiscovery/Random.hs [hunk 3]
```diff
@@ -76,8 +83,40 @@ addressToPath (Address addr) key = do
     payload <- deserialiseCbor decodeAddressPayload addr
     join $ deserialiseCbor (decodeAddressDerivationPath pwd) payload
 
-instance GenChange RndState where
-    genChange s = (error "GenChange RndState unimplemented", s)
+-- | Change state keeps track of the first free index to use and
+-- already used addresses (not only change addresses but ever known, ie. generated)
+data ChangeState = ChangeState
+    { forbidenAddresses :: Set Address
+    , passphrase :: Passphrase "encryption"
```
Comment (KtorZ) [inline lib/core/src/Cardano/Wallet/Primitive/AddressDiscovery/Random.hs]: Not obvious to me why we would need the encryption passphrase here?

File: lib/core/src/Cardano/Wallet/Primitive/AddressDiscovery/Random.hs [hunk 4]
```diff
@@ -76,8 +83,40 @@ addressToPath (Address addr) key = do
...
+emptyChangeState :: ChangeState
+emptyChangeState = ChangeState Set.empty (Passphrase "") (minBound, minBound)
+
+incrementIndices :: Word32 -> Word32 -> (Word32, Word32)
+incrementIndices accIx addrIx =
+    if (addrIx == maxBound && accIx == maxBound) then
+        (minBound, minBound)
+    else if (addrIx == maxBound) then
+        (accIx + 1, minBound)
+    else (accIx, addrIx + 1)
```
Comment (KtorZ) [inline lib/core/src/Cardano/Wallet/Primitive/AddressDiscovery/Random.hs]: Hmmm. To be honest, I'd rather go for the "random" approach first and see what it looks like, and then, see whether we have to come up with something clever (as hinted in the second part of the "decision" in the ticket) or not.

File: lib/core/src/Cardano/Wallet/Primitive/AddressDiscovery/Sequential.hs [hunk 5]
```diff
@@ -417,7 +417,7 @@ instance KeyToAddress t SeqKey => IsOurs (SeqState t) where
         in
             (ixs' `deepseq` ours `deepseq` ours, SeqState s1' s2' ixs')
 
-instance KeyToAddress t SeqKey => GenChange (SeqState t) where
+instance KeyToAddress t SeqKey => GenChange t (SeqState t) where
```
Comment (KtorZ) [inline lib/core/src/Cardano/Wallet/Primitive/AddressDiscovery/Sequential.hs]: This addition of the `t` parameter seems unnecessary. The generation of change shouldn't actually depend on the underlying target (apart from the address format, which is already captured by `KeyToAddress`)

File: lib/core/src/Cardano/Wallet/DB/Sqlite.hs (hunk 1 above)
Comment (paweljakubas) [reply lib/core/src/Cardano/Wallet/DB/Sqlite.hs]: indeed, I want to take care of DB in the last step

File: lib/core/src/Cardano/Wallet/Primitive/AddressDiscovery/Random.hs (hunk 3 above)
Comment (paweljakubas) [reply lib/core/src/Cardano/Wallet/Primitive/AddressDiscovery/Random.hs]: I use first : 
```
deriveAccountPrivateKey
    :: Passphrase "encryption"
    -> RndKey 'RootK XPrv
    -> Index 'Hardened 'AccountK
    -> RndKey 'AccountK XPrv
```
and then
```
deriveAddressPrivateKey
    :: Passphrase "encryption"
    -> RndKey 'AccountK XPrv
    -> Index 'Hardened 'AddressK
    -> RndKey 'AddressK XPrv
```
to derive address private key, which is used as an input to `keyToAddress`
I need `Passphrase "encryption"` here, right? Or I can just use for change addresses `Passphrase ""` ?

File: lib/core/src/Cardano/Wallet/Primitive/AddressDiscovery/Random.hs (hunk 4 above)
Comment (paweljakubas) [reply lib/core/src/Cardano/Wallet/Primitive/AddressDiscovery/Random.hs]: random, meaning pick at random two indices (accIx, addrIx), each between (minBound, maxBound) and check if present address constructed using them is in forbiddenAddress, right?

File: lib/core/src/Cardano/Wallet/Primitive/AddressDiscovery/Random.hs (hunk 3 above)
Comment (paweljakubas) [reply lib/core/src/Cardano/Wallet/Primitive/AddressDiscovery/Random.hs]: Or maybe I should follow strategy like in sequential mode : 
```
changeXPub <- deriveXPub DerivationScheme2 accXPub accIx 
addrXPub <- deriveXPub DerivationScheme2 changeXPub addrIx
```
But then I need `accXPub` which I need to setup when initiating emptyChange? If so, then I am not working with `'Hardened` but `'Soft` I guess

File: lib/core/src/Cardano/Wallet/Primitive/AddressDiscovery/Random.hs (hunk 4 above)
Comment (KtorZ) [reply lib/core/src/Cardano/Wallet/Primitive/AddressDiscovery/Random.hs]: Indeed. `forbiddenAddress` here really just being previously generated addresses, a.k.a known addresses.

File: lib/core/src/Cardano/Wallet/Primitive/AddressDiscovery/Random.hs (hunk 3 above)
Comment (KtorZ) [reply lib/core/src/Cardano/Wallet/Primitive/AddressDiscovery/Random.hs]: Hmmm. Okay. I'd suggest to move the `passphrase` from the `ChangeState` to `RndState` though. Or maybe, flatten `ChangeState` altogether in `RndState`. I don't the extra nesting buys us much :thinking:  ?

File: lib/core/src/Cardano/Wallet/Primitive/AddressDiscovery/Sequential.hs (hunk 5 above)
Comment (paweljakubas) [reply lib/core/src/Cardano/Wallet/Primitive/AddressDiscovery/Sequential.hs]: cannot make unit tests aware of target if t parameter is not present. Left with t parameter at this moment

File: lib/core/src/Cardano/Wallet/Primitive/AddressDiscovery/Random.hs (hunk 2 above)
Comment (paweljakubas) [reply lib/core/src/Cardano/Wallet/Primitive/AddressDiscovery/Random.hs]: done

File: lib/core/src/Cardano/Wallet/Primitive/AddressDiscovery/Random.hs (hunk 3 above)
Comment (paweljakubas) [reply lib/core/src/Cardano/Wallet/Primitive/AddressDiscovery/Random.hs]: removed `ChangeState` and put everything flatten in `RndState`

**Comment by paweljakubas** [comment]
hey @rvl - I have almost completed this PR. What is left is to do:  
1. replace  `incrementIndices` in `lib/core/src/Cardano/Wallet/Primitive/AddressDiscovery/Random.hs` in favor of randomly picking up `(accIx, addrIx)`. If so, then we can simplify `RndState` by removing `nextIndex` (see Matthias comment above : https://github.com/input-output-hk/cardano-wallet/pull/625#discussion_r313793847 )
2. Support DB storing of all fields in new `RndState`

Feel free to review, polish and merge this PR :-) Cheers!

**Comment by rvl** [comment]
Hi @paweljakubas 
1. I have changed it to select the address index at random and added the discovered and change addresses to the state. For now I did not randomly generate the account index because this would be a significant and unnecessary difference from the previous implementation.
2. I have added the SQLite tables and implemented `instance PersistState RndState`, but no tests yet. It's not strictly necessary to save the `StdGen` to the database, but just easier like that.

File: lib/core/src/Cardano/Wallet/Primitive/AddressDiscovery/Sequential.hs (hunk 5 above)
Comment (rvl) [reply lib/core/src/Cardano/Wallet/Primitive/AddressDiscovery/Sequential.hs]: I found that I could remove `t` from `GenChange` if I added it to `RndState`.

**Review by KtorZ — APPROVED** [review]

**Review by rvl — APPROVED** [review]

**Pull request merged**